### PR TITLE
Add a confirmation dialog to config action buttons.

### LINKED ISSUE #850: Cancel button on config panel does not close the panel
- Go to /models/cmr?entity=logstash&panel=config
- Click 'Cancel'
- Panel does not close

### PR DESCRIPTION
**Pull request by hatched**
## Done

Add a confirmation dialog to the configuration action buttons [cancel, save]

Note: use w=1 for review.

## QA

Use the demo link and the model `charmed-kubernetes` with the app `containerd` then branch out to try other models/apps to chaos monkey it.
- Without editing any config, clicking the `X` or `Cancel` should just close the panel.
- If you edit a config and click `Cancel` it should confirm, try both buttons, Also do the same with the `X`
- Now edit a config and click save, it should confirm, try both buttons.
- Chaos monkey!

## Details

Fixes #850 
Fixes https://github.com/canonical-web-and-design/juju-squad/issues/1581

## Screenshots

Screen Shot 2021-02-02 at 3 39 13 PM
Screen Shot 2021-02-02 at 3 39 24 PM

### CODE REVIEW

File: src/components/SlidePanel/SlidePanel.js [hunk 1]
```diff
@@ -1,23 +1,10 @@
-import { cloneElement } from "react";
 import classnames from "classnames";
 import Spinner from "@canonical/react-components/dist/components/Spinner";
 
 import useEventListener from "hooks/useEventListener";
 
 import "./_slide-panel.scss";
 
-function generatePanelContent(isLoading, children, onClose) {
```
Comment (hatched) [inline src/components/SlidePanel/SlidePanel.js:9]: The usage of `SlidePanel` is going to change from being a wrapper of child content to being included as part of that child content itself. This allows much easier control over what elements respond to the 'click outside' and the ability to interact with the various closure mechanisms of the panel.

File: src/index.js [hunk 2]
```diff
@@ -74,11 +74,6 @@ function bootstrap() {
       juju: jujuReducer,
       ui: uiReducer,
     }),
-    {
-      ui: {
```
Comment (hatched) [inline src/index.js:78]: This default is now defined in the reducer as is the recommended pattern.

File: src/components/ConfirmationModal/_confirmation-modal.scss [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+      background-color: inherit;
+      border: none;
+      padding: 0;
+      // extra styles
+      white-space: pre-wrap;
+      word-break: keep-all;
+    }
+
+    &__dialog {
+      max-width: 522px;
```
Comment (barrymcgee) [inline src/components/ConfirmationModal/_confirmation-modal.scss:25]: What is this magic number?

File: src/panels/ConfigPanel/ConfigPanel.tsx [hunk 4]
```diff
@@ -48,21 +50,25 @@ export type ConfigProps = {
...
-  const [enableSave, setEnableSave] = useState<Boolean>(false);
-  const [showResetAll, setShowResetAll] = useState<Boolean>(false);
-  const [isLoading, setIsLoading] = useState<Boolean>(true);
-  const [savingConfig, setSavingConfig] = useState<Boolean>(false);
+  const [enableSave, setEnableSave] = useState<boolean>(false);
+  const [showResetAll, setShowResetAll] = useState<boolean>(false);
+  const [isLoading, setIsLoading] = useState<boolean>(true);
+  const [savingConfig, setSavingConfig] = useState<boolean>(false);
+  const [confirmOpen, setConfirmOpen] = useState<boolean>(false);
+  const [confirmType, setConfirmType] = useState<ConfirmTypes>("apply");
```
Comment (barrymcgee) [inline src/panels/ConfigPanel/ConfigPanel.tsx:71]: Merging these values into one 'confirm' object and updating it once would reduce rerendering and eliminate possible race conditons.

File: src/panels/ConfigPanel/ConfigPanel.tsx [hunk 5]
```diff
@@ -300,6 +381,25 @@ function generateConfigElementList(
   return elements;
 }
 
+function generateChangedKeyValues(config: Config) {
+  const changedValues = Object.keys(config).reduce(
+    (acc: ReactNode[], key: string) => {
```
Comment (barrymcgee) [inline src/panels/ConfigPanel/ConfigPanel.tsx:386]: It's unclear to me what `acc` means here? Can you use a clearer name?

File: src/panels/ConfigPanel/_config-panel.scss [hunk 6]
```diff
@@ -131,24 +143,35 @@
 
   &__drawer {
     background-color: white;
-    border-top: 1px solid $color-mid-light;
     bottom: 0;
-    max-height: 130px;
-    padding-right: 1.5rem;
-    padding-top: 1rem;
+    box-shadow: none;
+    max-height: 500px; // xxx fix me
```
Comment (barrymcgee) [inline src/panels/ConfigPanel/_config-panel.scss:148]: Fix what?

File: src/ui/actions.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
+// Action labels
+export enum actionsList {
+  userMenuActive = "TOGGLE_USER_MENU",
+  confirmationModalActive = "TOGGLE_CONFIRMATION_Modal",
```
Comment (barrymcgee) [inline src/ui/actions.ts:4]: Why the change of case?

**Review by barrymcgee — CHANGES_REQUESTED** [review]

**Comment by barrymcgee** [comment]
Modal is appearing bottom left?

Screenshot 2021-02-03 at 13 48 51

**Comment by barrymcgee** [comment]
- [ ] Form is not accessible via keyboard.. make a change, tab to 'Apply changes'.. focus then switches to window
- [ ] Form should cancel when Esc key is pressed

**Comment by barrymcgee** [comment]
Small screen view is broken.

Screenshot 2021-02-03 at 14 10 52

**Comment by long-chung** [comment]
Can't find any other things other than what Barry has flagged. -1 this until its resolved.

**Comment by long-chung** [comment]
Also i'm not seeing any descriptions of form parameters in mobile

File: src/components/ConfirmationModal/_confirmation-modal.scss (hunk 3 above)
Comment (hatched) [reply src/components/ConfirmationModal/_confirmation-modal.scss:25]: Heh yes pretty magical, it was the max-width set in the design :)

File: src/panels/ConfigPanel/ConfigPanel.tsx (hunk 5 above)
Comment (hatched) [reply src/panels/ConfigPanel/ConfigPanel.tsx:386]: ahh, yes in the `reduce` call it's the `accumulator`. I can expand the variable name.

File: src/panels/ConfigPanel/_config-panel.scss (hunk 6 above)
Comment (hatched) [reply src/panels/ConfigPanel/_config-panel.scss:148]: Heh good catch, this was to be removed.

File: src/ui/actions.ts (hunk 7 above)
Comment (hatched) [reply src/ui/actions.ts:4]: typo :D

File: src/panels/ConfigPanel/ConfigPanel.tsx (hunk 4 above)
Comment (hatched) [reply src/panels/ConfigPanel/ConfigPanel.tsx:71]: Good call.

**Comment by hatched** [comment]
@barrymcgee @long-chung all comments have been addressed (and some bonus fixes too) plz take another look.

@long-chung they appear to be working here across browsers..maybe try a hard refresh.

Screen Shot 2021-02-03 at 3 15 51 PM

File: src/components/ConfirmationModal/ConfirmationModal.tsx [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+  const portalHost =
+    document.querySelector("#confirmation-modal-container") || document.body;
+
+  useEffect(() => {
+    if (modalRef.current) {
+      const neutralButton: HTMLButtonElement | null = modalRef?.current?.querySelector(
+        "button.p-button--neutral"
+      );
+      // If the modal has a neutral button then focus on that after opening
+      // to make a better keyboard navigatione experience.
```
Comment (barrymcgee) [inline src/components/ConfirmationModal/ConfirmationModal.tsx:27]:
```suggestion
      // to make a better keyboard navigation experience.
```

**Review by barrymcgee — APPROVED** [review]
